java-support/src/main/scala/io/cloudstate/javasupport/impl/controller/AnnotationBasedControllerSupport.scala

```diff
@@ -0,0 +1,462 @@
+          throw new RuntimeException(
+            s"Incompatible message class $messageType for call $method, expected ${resolvedType.typeClass}"
+          )
+        } else { envelope =>
+          MessageEnvelope.of(
+            resolvedType.parseFrom(envelope.payload.getValue).asInstanceOf[AnyRef],
+            envelope.metadata
+          )
+        }
+      case payload =>
+        if (payload != resolvedType.typeClass) {
+          throw new RuntimeException(
+            s"Incompatible message class $payload for call $method, expected ${resolvedType.typeClass}"
+          )
+        } else { envelope =>
+          resolvedType.parseFrom(envelope.payload.getValue).asInstanceOf[AnyRef]
+        }
+    }
+}
+
+private class PayloadParameterHandler(mapper: MessageEnvelope[JavaPbAny] => AnyRef)
+    extends ParameterHandler[MessageEnvelope[JavaPbAny], ControllerContext] {
+  override def apply(ctx: InvocationContext[MessageEnvelope[JavaPbAny], ControllerContext]): AnyRef =
+    mapper(ctx.mainArgument)
+}
+
+private class StreamedPayloadParameterHandler(mapper: javadsl.Source[MessageEnvelope[JavaPbAny], NotUsed] => AnyRef)
+    extends ParameterHandler[javadsl.Source[MessageEnvelope[JavaPbAny], NotUsed], ControllerContext] {
+  override def apply(
+      ctx: InvocationContext[javadsl.Source[MessageEnvelope[JavaPbAny], NotUsed], ControllerContext]
+  ): AnyRef =
+    mapper(ctx.mainArgument)
+}
+
+private trait UnaryInSupport {
+  protected val method: Method
+  protected val serviceMethod: ResolvedServiceMethod[_, _]
+
+  protected val parameters: Array[ParameterHandler[MessageEnvelope[JavaPbAny], ControllerContext]] =
+    ReflectionHelper.getParameterHandlers[MessageEnvelope[JavaPbAny], ControllerContext](method) {
+      case payload =>
+        new PayloadParameterHandler(
+          ControllerReflection
+            .getInputParameterMapper(serviceMethod.name, serviceMethod.inputType, payload.genericParameterType)
+        )
+    }
+}
+
+private trait UnaryOutSupport {
+  protected val method: Method
+  protected val serviceMethod: ResolvedServiceMethod[_, _]
+
+  protected val outputMapper: Any => CompletionStage[ControllerReply[JavaPbAny]] = method.getReturnType match {
+    case cstage if cstage == classOf[CompletionStage[_]] =>
+      val cstageType = ReflectionHelper.getGenericFirstParameter(method.getGenericReturnType)
+      val mapper =
+        ControllerReflection.getOutputParameterMapper(serviceMethod.name, serviceMethod.outputType, cstageType)
+
+      any: Any => any.asInstanceOf[CompletionStage[Any]].thenApply(mapper.apply)
+    case _ =>
+      val mapper = ControllerReflection.getOutputParameterMapper(serviceMethod.name,
+                                                                 serviceMethod.outputType,
+                                                                 method.getGenericReturnType)
+
+      any: Any => CompletableFuture.completedFuture(mapper(any))
+  }
+}
+
+private trait StreamedInSupport {
```

Comment:
it would be nice if we can support some kind of java.util.collection for streamedin. we had the disccusion here - https://github.com/cloudstateio/cloudstate/issues/266. what do you think @jroper?

Reply:
We could definitely support this, though my question would be, what's the use case? If you want to handle messages that you know will fit in memory, why not use a unary call and put those messages in a repeated field in the protobuf? My view on streamed calls is that they are useful for the following cases:

* Where you don't want to buffer all the elements in memory because that could cause you to run out of memory.
* Where you want to handle elements as they arrive, eg, in a chat application, each message should be handled immediately.

In both of the above cases, using a collection would defeat the purpose of using a streamed call - in fact, a call that handles all its messages in a single collection is by definition not streamed - so why use a streamed call for something that isn't streamed? Supporting this seems to be promoting bad practices to me.

Reply:
There is no special use case for supporting collection for streamed-in calls. it is an anti pattern to use streamed calls to buffer all the in memory. the unary call can be used and it is fine for me. So i agree with your explanation.

Reply:
Hi @jroper There are use cases for batch message handling, imagine a stream processing system where some other system has added messages that belong to some business category and should be aggregated into lists, very common in ETL processing, Machine Learning, CEP, and etc. ... In these cases the intake system will send finite data in batches that can be processed as a collection. Bearing in mind that a StreamIn will always be finite because it is expected to have a unary output, so it makes sense to support Collections, even though the same use case can be achieved with reactive flow operators

Reply:
@ralphlaude It's not an anti pattern, it's just something you didn't think of

Reply:
Just remembering that I am not referring to "Streamed" calls as Guy mentioned, I am referring to the only and exclusively "StreamIn" calls, that is, finite incoming flows. And yes, it may be that the memory is not enough, but on the other hand the user would obviously be aware of this when making the choice of the desired operator in his method signature.

Reply:
Stream in is not necessarily finite. It might ordinarily be infinite, but the receiver may change state, to a state where it can no longer accept messages, and in that case the out message produced would indicate information about that, for example, it might indicate that a long lived entity has now been deleted, or has been moved and the message may indicate how to reconnect, etc.

It might also be finite, but messages might need to be processed in real time, for example, you might use a stream in to report to a server the progress of a user integration, and once done, the server sends a single acknowledgement message. The thing happening has a finite number of state changes, so you're going to send a finite number of messages, but you want them handled as they arrived, not batched then handled.

If I was sending finite data to be processed in batches, I'd still rather model that as a single message containing the batch of data to process - this forces the sender to think about how much data can reasonably be fit into a single batch. Of course, that single message could still be very large and may cause things downstream to run out of memory, but it makes the contract clear, if the sender sends 100mb in a single message, that's obviously bad, but it's not obviously bad that sending 100mb spread across a stream is bad. If I did decide to use a stream, and I wanted downstream to process in batches, then I'd rather model that in the receiver using a grouped streaming operator, so that the receiver does do batching, but limits the amount that it buffers, and may process big batches as multiple sub batches, now the sender doesn't have to have any knowledge of the resource constraints of the receiver.

The advantage of supporting the collection types is convenience for the simple case, but increased danger - if we support collection types, then we have to document them (otherwise they may as well not exist), and many users will copy and paste those code examples without considering the consequences. The advantage of not supporting them is that we force the user to think about the fact that this is a stream and could cause them to run out of resources - they can still buffer them all then process them, but they've had to explicitly do that in code, and then in our documentation, we won't have any examples of buffering the whole stream, if we want to document batch handling, we can show how to do it safely with bounded buffer groups, so the users that just copy and paste will build something that by default is safe.